backend/src/test/java/com/google/sps/EmailNotifierTest.java

```diff
@@ -0,0 +1,159 @@
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public class EmailNotifierTest {
+
+  private EmailNotifier emailNotifier;
+  private Gmail gmail;
+  private Gmail.Users users;
+  private Gmail.Users.Messages messages;
+  private Message testMessage;
+
+  @Before
+  public void setUp() throws MessagingException, IOException {
+    gmail = mock(Gmail.class);
+    // Mock to return list of users
+    users = mock(Gmail.Users.class);
+    // Mock to return list of messages.
+    messages = mock(Gmail.Users.Messages.class);
+    MimeMessage mimemessage = createTestMessage();
+    testMessage = createMessageFromMime(mimemessage);
+    // Test Instance of email notifier class.
+    emailNotifier = new EmailNotifier("John", "jdoe@gmail.com", gmail);
+    // Emulating real behavior due to when method, will return list of users.
+    when(gmail.users()).thenReturn(users);
+    // Emulating real behaviour due to when method, will return list of messages
+    when(users.messages()).thenReturn(messages);
+  }
+
+  public MimeMessage createTestMessage() throws MessagingException {
+    Properties properties = System.getProperties();
+    Session session = Session.getDefaultInstance(properties);
+    MimeMessage message = new MimeMessage(session);
+    Address[] addresses =
+        new Address[] {
+          new InternetAddress("tcwang@google.com"),
+          new InternetAddress("grantjustice@google.com"),
+          new InternetAddress("kevinhowald@google.com")
+        };
+    message.setFrom(new InternetAddress("Adlib-Step@gmail.com"));
+    message.setSubject("Test Subject");
+    message.setText("Test Text");
+    message.setRecipients(javax.mail.Message.RecipientType.TO, addresses);
+    return message;
+  }
+
+  public Message createMessageFromMime(MimeMessage mimeMessage)
+      throws IOException, MessagingException {
+    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+    mimeMessage.writeTo(buffer);
+    byte[] bytes = buffer.toByteArray();
+    String encodedEmail = Base64.encodeBase64URLSafeString(bytes);
+    Message message = new Message();
+    message.setRaw(encodedEmail);
+    return message;
+  }
+
+  public MimeMessage convertToMimeMessage(Message msg) throws MessagingException {
+    Properties properties = System.getProperties();
+    Session session = Session.getDefaultInstance(properties);
+    byte[] bytes = msg.decodeRaw();
+    InputStream targetStream = new ByteArrayInputStream(bytes);
+    return new MimeMessage(session, targetStream);
+  }
+
+  @Test
+  public void constructorShouldSetNameAndEmailTo() {
+    assertEquals("John", emailNotifier.getRecipientName());
+    assertEquals("jdoe@gmail.com", emailNotifier.getToEmail());
+  }
+
+  @Test
+  public void testMessageHasCorrectApplicationName() throws MessagingException, IOException {
+    ArgumentCaptor<Message> argument = ArgumentCaptor.forClass(Message.class);
+    emailNotifier.getService().users().messages().send("anyString", testMessage);
```

Comment:
If the test accesses the gmail service directly and sends the message then we're missing out on testing the code within the EmailNotifier class.

Please make sure that this test and all the others are calling notifyUser.

Reply:
i believe I have addressed these issues via my understanding of it !